### ISSUE TITLE
Feature request for Edge Futures: #map and #flat_map

### ISSUE THREAD

**Issue by ms-ati** [ruby-concurrency/concurrent-ruby#614]
Hi all,

Would you consider accepting a pull request to add `#map` and `#flat_map` to the Edge Futures?

I think it would make functional composition easier and more clear, especially after #611 is fixed.

For example:

```ruby
# For all Edge Futures...

alias_method :map, :then

def flat_map(&block)
  map(&block).flat
end
```

So then we could write:

```ruby
f = Concurrent.delay { 42 }.
  map { |n| n / 7 }.
  flat_map { |n| Concurrent.zip(Array.new(n) { |i| Concurrent.delay { i } }) }.
  map { |*values| values.inject(&:+) }

# still nothing has happened

f.value!

# => 15
```

What do you all think?

**Comment by pitr-ch** [comment]
Thank you for your interest in edge promises. I would like to avoid creating aliases, the promise implementation follows js naming so we would like to stick it. I think `then_flat` is acceptable (there are other `then_*` methods like `then_ask`, `then_select`, etc.). We'll appreciate a PR! However there is big renaming of Promise API awaiting in a PR to be merged after 1.0.3 release. You could either work against https://github.com/ruby-concurrency/concurrent-ruby/pull/522 or wait until it's merged into master.

**Comment by ms-ati** [comment]
Ok -- what are you guys thinking about `.sequence` and `.traverse` for example? Best live somewhere else?

**Comment by pitr-ch** [comment]
could you link what sequence and traverse do you mean?

**Comment by pitr-ch** [comment]
@ms-ati did you had following in mind?
```ruby
require 'concurrent-edge'
Promises = Concurrent::Promises
resources = Array.new(10) { |i| i }
r = resources.reduce(Promises.fulfilled_future([])) do |f, resource|
  f.then(resource) { |arr, resource| [*arr, resource**2] }
end.value!
p r # => [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

```

**Comment by pitr-ch** [comment]
Aliases added in https://github.com/ruby-concurrency/concurrent-ruby/pull/637/files#diff-4563ba78b4a06ead2e72d5d5cbe03ac3R20, closing.
